enforce_rules.test.js

```diff
@@ -0,0 +1,49 @@
+let {allowMessage, rule1Evaluation, rule4Evaluation} = require("./enforce-rules");
+let test_data = require("./enforce_rules_test_data.json");
+
+// RULE1EVALUATION
```

Comment:
One of the really nifty things that jest (and other RSpec inspired test runners) provides is the [`describe`](https://jestjs.io/docs/en/api#describename-fn) directive which allows for arbitrary nesting of related testing, which provides neat things with scoping of variables, grouping tests together to make it easy to run just a particular grouping via `--grep`, etc.

In this case, we would probably want to have a single level of grouping which is the rule that's being tested, and so you would have:

```javascript
describe('rule1Evaluation', () => {
  // your tests
});
```

Next, looking at the contents of `enforce_rules_test_data.json`, the JSON object is more or less the same in all cases, with just the difference being in the `bid` key/value pair. Well, with the above grouping, we can use that to create a nice scoped variable to contain the "shell test data", and then within each test add in the bid key/value pair we want.
This allows us to avoid having a JSON file with a lot of repeated data, making it easier to add new tests and such if
need be.

An example of how this might look would be:

```javascript
describe('rule1Evaluation', () => {
  const data = {
    roundId: "true",
    speaker: "Human",
    addressee: "Watson",
    text: "Watson, I will buy 3 eggs for $4",
    role: "buyer",
    timestamp: 1595945030581,
  };

  test('accept pass', () => {
    const testData = Object.assign({}, data, {bid: {"type": "Accept", "price": {"value": 10}}});
    expect(rule1Evaluation(testData, 50)).toEqual({permit: true, rationale: null});
  });

  test('non-accept pass', () => {
    const testData = Object.assign({}, data, {bid: {"type": "Reject", "price": {"value": 10}}});
    expect(rule1Evaluation(testData, 50)).toEqual({permit: true, rationale: null});
  });

  // rest of the tests
});
```

Alright, with that done, we now see that the difference between tests lie then in:
1. what is `bid` key/value pair equal to
2. what is the value of the human budget (second parameter to `rule1Evaluation`)
3. expected result

but that the actual structure of the test is the same in each case which is:
1. Do `Object.assign` to generate the `testData` object
2. Run `expect(rule1Evaluation(..., ...)).toEqual(....);`

Well, we can then utilize the concept of "Parameterized Tests", which the basics of it being
that it allows us to write a test once, and then test it using a set of different inputs and
outputs. This blog post gives an alright overview of the concept, and doing it in jest:
https://itnext.io/reduce-unit-tests-boilerplate-with-jests-each-syntax-f5e48828437f

Now, with that in-hand, we can rewrite the above to be:

```javascript
describe('rule1Evaluation', () => {
  const data = {
    roundId: "true",
    speaker: "Human",
    addressee: "Watson",
    text: "Watson, I will buy 3 eggs for $4",
    role: "buyer",
    timestamp: 1595945030581,
  };

  test.each([
    ['accept pass', {"type": "Accept", "price": {"value": 10}}, 50, {permit: true, rationale: null}],
    ['non-accept pass', {"type": "Reject", "price": {"value": 10}}, 50, {permit: true, rationale: null}],
    // rest of test parameters
  ]).('%s', (testName, bid, budget, expected) => {
    const testData = Object.assign({}, data, {bid});
    expect(rule1Evaluation(testData, budget)).toEqual(expected);
  });
});
```

This now gets us a nice succient, and easy to add to test for `rule1Evaluation`. 😃